### PR TITLE
EventGrid: PowerShell cmdlets for API Version 2019-01-01

### PR DESCRIPTION
**Pull request by ahamad-MS**
## Description

Powershell work for the new 2019-01-01 API version. This version adds Powershell support for the following features:
    a.	Deadletter destination.
    b.	Storage queue as destination.
    c.	Hybrid connection as destination.
    d.	Retry policy.
    e.	Manual handshake.

These features were initially introduced in 2018-05-01-preview API Version. Full description of these features can be found here: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/53 

Other features that are still in preview and can still be accessed using the API version 2018-09-15-preview are:
    a.	Domains.
    b.	Advanced filters support.
    c.	Event subscription expiration date
    d.	Input mapping and event delivery schema.

Here is a summary of the updated cmdlets for 2019-01-01 API version.

* Update the following cmdlets:
	- New-AzureRmEventGridSubscription: Add new optional parameters for specifying:
		- Event Time-To-Live,
		- Maximum number of delivery attempts for the events,
		- Dead letter endpoint.
	- Update-AzureRmEventGridSubscription: Add new optional parameters for specifying:
		- Event Time-To-Live,
		- Maximum number of delivery attempts for the events,
		- Dead letter endpoint.

* Add new enum values (namely, storageQueue and hybridConnection) for EndpointType option in New-AzureRmEventGridSubscription and Update-AzureRmEventGridSubscription cmdlets.

* Show warning message if creating or updating the event subscription is expected to entail manual action from user.

## Checklist

- [X] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [X] The title of the PR is clear and informative
- [X] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [X] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [X] If applicable, the changes made in the PR have proper test coverage
- [X] For public API changes to cmdlets:
    - [X] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)

Link to design review for 2018-05-01-preview when these features were introduced: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/53

    - [X] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by maddieclayton** [comment]
@ahamad-MS Please resolve merge conflicts

**Comment by ahamad-MS** [comment]
@maddieclayton  Can you please help with this PR. Thanks

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs [hunk 1]
```diff
@@ -22,87 +22,94 @@
...
+            Mandatory = true,
             ValueFromPipeline = true,
             Position = 0,
-            HelpMessage = "EventGrid Topic object.",
-            ParameterSetName = EventSubscriptionInputObjectParameterSet)]
+            HelpMessage = EventGridConstants.TopicInputObjectHelp,
+            ParameterSetName = EventSubscriptionCustomTopicInputObjectParameterSet)]
         [ValidateNotNullOrEmpty]
-        public PSTopic InputObject { get; set; }
+        public PSTopic CustomTopicInputObject { get; set; }
```
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs]: We simply changed the name of the input object which is used in piping. there is no explicit use of this name in the user input commands. We needed to change this to make it more appropriate as we added domain and domain topic resources which have input object definition too. There is no new change here as this is from the previous GA work. However, even though we don't consider this as blocking change, I am reverting these names back to InputObject to remove any concern.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs [hunk 2]
```diff
@@ -20,50 +22,57 @@
...
+            Mandatory = true,
             ValueFromPipeline = true,
             Position = 0,
-            HelpMessage = EventGridConstants.TopicInputObject,
-            ParameterSetName = EventSubscriptionInputObjectParameterSet)]
+            HelpMessage = EventGridConstants.TopicInputObjectHelp,
+            ParameterSetName = EventSubscriptionCustomTopicInputObjectParameterSet)]
         [ValidateNotNullOrEmpty]
-        public PSTopic InputObject { get; set; }
+        public PSTopic CustomTopicInputObject { get; set; }
```
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: Retaining old name even though we don't see this as a breaking change as this is not user specified input.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs [hunk 3]
```diff
@@ -81,7 +90,7 @@ public class NewAzureEventGridSubscription : AzureEventGridCmdletBase
             ParameterSetName = CustomTopicEventSubscriptionParameterSet)]
         [Parameter(
             Mandatory = true,
-            ValueFromPipelineByPropertyName = true,
+            ValueFromPipelineByPropertyName = false,
```
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: Won't Fix:

This is part of commit 4efd0aa50bcd6f6c11433dca98f6f42e824aee56 which I believe wad addressing a specific bug. 

https://github.com/Azure/azure-powershell/commit/4efd0aa50bcd6f6c11433dca98f6f42e824aee56#diff-b15ea754456daa297b8be0d8d2e89c7a

@kalyanaj  I don't seem to remember the issue you addressed here and could not find the issue itself in GitHub. Can you please help addressing this concern. thanks.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs [hunk 4]
```diff
@@ -22,87 +22,94 @@
...
+        DefaultParameterSetName = EventSubscriptionTopicNameParameterSet),
+    OutputType(typeof(PSEventSubscription))]
+
     public class GetAzureRmEventGridSubscription : AzureEventGridCmdletBase
     {
-        [Parameter(Mandatory = false,
+        [Parameter(
+            Mandatory = false,
             ValueFromPipelineByPropertyName = true,
-            Position = 0,
```
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs]: Won't fix:
We were asked to remove any position for any new parameters that are optional. Please see the comments in this PR. We apply same for all optional parameters in the code for consistency.

https://github.com/Azure/azure-powershell/pull/6238

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs [hunk 5]
```diff
@@ -90,21 +99,20 @@ public class NewAzureEventGridSubscription : AzureEventGridCmdletBase
...
             Mandatory = true,
             ValueFromPipelineByPropertyName = true,
             Position = 2,
-            HelpMessage = "The resource group of the topic.",
+            HelpMessage = EventGridConstants.ResourceGroupNameHelp,
             ParameterSetName = CustomTopicEventSubscriptionParameterSet)]
         [Parameter(
             Mandatory = false,
             ValueFromPipelineByPropertyName = true,
-            Position = 2,
```
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: same as above.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/UpdateAzureEventGridSubscription.cs [hunk 6]
```diff
@@ -101,10 +109,10 @@ public class UpdateAzureEventGridSubscription : AzureEventGridCmdletBase
         [Parameter(
             Mandatory = false,
             HelpMessage = EventGridConstants.EndpointTypeHelp,
-            ParameterSetName = EventSubscriptionInputObjectParameterSet)]
-        [ValidateSet("webhook", "eventhub", IgnoreCase = true)]
+            ParameterSetName = EventSubscriptionCustomTopicInputObjectParameterSet)]
+        [ValidateSet(EventGridConstants.Webhook, EventGridConstants.EventHub, EventGridConstants.StorageQueue, EventGridConstants.HybridConnection, IgnoreCase = true)]
         [ValidateNotNullOrEmpty]
-        public string EndpointType { get; set; }
+        public string EndpointType { get; set; } = EventGridConstants.Webhook;
```
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/UpdateAzureEventGridSubscription.cs]: It is assumed to be same Eventridconstants.Webhook. Please refer to this check in line 231 in src\ResourceManager\EventGrid\Commands.EventGrid\Utilities\EventGridClient.cs.

We need explicit value here as we added additional warning message based on the endpoint type.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs [hunk 7]
```diff
@@ -217,92 +213,174 @@ public class NewAzureEventGridSubscription : AzureEventGridCmdletBase
...
+                    // ResourceID not specified, build the scope for the event subscription for either the
                     // subscription, or resource group, or custom topic depending on which of the parameters are provided.
+
                     scope = EventGridUtils.GetScope(this.DefaultContext.Subscription.Id, this.ResourceGroupName, this.TopicName);
                 }
 
-                EventSubscription eventSubscription = this.Client.CreateEventSubscription(scope,
-                    this.EventSubscriptionName, this.Endpoint, this.EndpointType, this.SubjectBeginsWith,
-                    this.SubjectEndsWith, isSubjectCaseSensitive, this.IncludedEventType, this.Label);
+                if (this.MaxDeliveryAttempt.HasValue || this.EventTtl.HasValue)
```
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: thanks for the suggestion. Won't fix. Please check the explanation below

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs]: `InputObject` is an explicit parameter. You can see it in the 4th parameter set in the cmdlet:
https://docs.microsoft.com/en-us/powershell/module/az.eventgrid/get-azeventgridsubscription?view=azps-1.0.0

A valid call to your cmdlet is:
```powershell
Get-AzEventGridSubscription -InputObject $topic
```

Also, it is _by convention_ that we use `InputObject` to describe the parent resource for Az cmdlets.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: Same comment as above. `InputObject` is an explicit parameter that can be called by name.
Third parameter set in these docs:
https://docs.microsoft.com/en-us/powershell/module/az.eventgrid/new-azeventgridsubscription?view=azps-1.0.0

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: Unfortunately, any changes to the calling pattern is a breaking change, and we cannot accept those changes outside of a breaking change release.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs]: Correct. You were asked to remove position from any **new** parameters. This is not a new parameter. It cannot be accepted outside of a breaking change release.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: Correct. You were asked to remove position from any **new** parameters. This is not a new parameter. It cannot be accepted outside of a breaking change release.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/UpdateAzureEventGridSubscription.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/UpdateAzureEventGridSubscription.cs]: Sounds good.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: As mentioned above, we don't accept nullable parameters outside of an Update cmdlet. This code above tells you how to determine if someone provided the parameter. In this case, if your service accepts `null`, then if the parameter was not bound (not provided), send `null` to your service.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: OK.. reverted. We will discuss off line on when we need this change to be in.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs]: OK. reverted.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: Fixed.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: OK, we are updating it to use int

**Comment by ahamad-MS** [comment]
@MiYanni  all comments are addressed Please take another look. thx

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: This has not been fixed.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs [hunk 8]
```diff
@@ -115,90 +123,78 @@ public class NewAzureEventGridSubscription : AzureEventGridCmdletBase
...
             ValueFromPipelineByPropertyName = false,
-            Position = 3,
             HelpMessage = EventGridConstants.EndpointTypeHelp,
-            ParameterSetName = EventSubscriptionInputObjectParameterSet)]
-        [ValidateSet("webhook", "eventhub", IgnoreCase = true)]
+            ParameterSetName = EventSubscriptionCustomTopicInputObjectParameterSet)]
+        [ValidateSet(EventGridConstants.Webhook, EventGridConstants.EventHub, EventGridConstants.StorageQueue, EventGridConstants.HybridConnection, IgnoreCase = true)]
         [ValidateNotNullOrEmpty]
-        public string EndpointType { get; set; }
+        public string EndpointType { get; set; } = EventGridConstants.Webhook;
```
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: This was discussed in another comment.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/UpdateAzureEventGridSubscription.cs [hunk 9]
```diff
@@ -195,10 +203,79 @@ public class UpdateAzureEventGridSubscription : AzureEventGridCmdletBase
...
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = false,
+            HelpMessage = EventGridConstants.EventTtlHelp,
+            ParameterSetName = ResourceIdEventSubscriptionParameterSet)]
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = false,
+            HelpMessage = EventGridConstants.EventTtlHelp,
+            ParameterSetName = EventSubscriptionCustomTopicInputObjectParameterSet)]
+        [ValidateRange(1, 1440)]
```
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/UpdateAzureEventGridSubscription.cs]: This was fixed.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/UpdateAzureEventGridSubscription.cs [hunk 10]
```diff
@@ -195,10 +203,79 @@ public class UpdateAzureEventGridSubscription : AzureEventGridCmdletBase
...
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = false,
+            HelpMessage = EventGridConstants.MaxDeliveryAttemptHelp,
+            ParameterSetName = ResourceIdEventSubscriptionParameterSet)]
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = false,
+            HelpMessage = EventGridConstants.MaxDeliveryAttemptHelp,
+            ParameterSetName = EventSubscriptionCustomTopicInputObjectParameterSet)]
+        [ValidateRange(1, 30)]
```
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/UpdateAzureEventGridSubscription.cs]: This was fixed.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/UpdateAzureEventGridSubscription.cs [hunk 11]
```diff
@@ -195,10 +203,79 @@ public class UpdateAzureEventGridSubscription : AzureEventGridCmdletBase
...
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = false,
+            HelpMessage = EventGridConstants.DeadletterEndpointHelp,
+            ParameterSetName = ResourceIdEventSubscriptionParameterSet)]
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = false,
+            HelpMessage = EventGridConstants.DeadletterEndpointHelp,
+            ParameterSetName = EventSubscriptionCustomTopicInputObjectParameterSet)]
+        [ValidateNotNullOrEmpty]
```
[the remark this answers is not in the document]
Comment (MiYanni) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/UpdateAzureEventGridSubscription.cs]: This makes sense with the rest of the parameters' validations.

File: src/ResourceManager/EventGrid/Commands.EventGrid/Utilities/EventGridUtils.cs [hunk 12]
```diff
@@ -47,7 +49,7 @@ public static string GetScope(string subscriptionId, string resourceGroupName, s
                 throw new ArgumentNullException(nameof(subscriptionId));
             }
 
-            string scope;
+            string scope = null;
```
Comment (MiYanni) [inline src/ResourceManager/EventGrid/Commands.EventGrid/Utilities/EventGridUtils.cs]: `string` is null at declaration (meaning `= null` does nothing). Just a comment; no change required.

File: src/ResourceManager/EventGrid/Commands.EventGrid/help/New-AzEventGridSubscription.md [hunk 13]
```diff
@@ -150,23 +181,23 @@ This can be webhook or eventhub
 Type: System.String
 Parameter Sets: ResourceGroupNameParameterSet, ResourceIdEventSubscriptionParameterSet, CustomTopicEventSubscriptionParameterSet
 Aliases:
-Accepted values: webhook, eventhub
+Accepted values: webhook, eventhub, storagequeue, hybridconnection
 
 Required: False
-Position: 4
+Position: Named
```
Comment (MiYanni) [inline src/ResourceManager/EventGrid/Commands.EventGrid/help/New-AzEventGridSubscription.md]: When the positions are re-added, the documentation should be regenerated, so this change is reverted.

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/GetAzureEventGridSubscription.cs]: fixed

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: fixed

File: src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/EventSubscription/NewAzureEventGridSubscription.cs]: fixed

File: src/ResourceManager/EventGrid/Commands.EventGrid/help/New-AzEventGridSubscription.md (hunk 13 above)
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/help/New-AzEventGridSubscription.md]: fixed

Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/help/New-AzEventGridSubscription.md]: @MiYanni all comments were addressed.

**Comment by MiYanni** [comment]
@ahamad-MS Since you changed the name of the parameter set, `EventSubscriptionCustomTopicInputObjectParameterSet` is flagged by static analysis for having too many positional parameters. Please add an exception for this.

**Pull request merged**